plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -221,13 +221,17 @@ private List<SchemaTableName> listTablesInRemoteSchema(BigQueryClient client, St
                                 () -> tableNames.add(new SchemaTableName(client.toSchemaName(DatasetId.of(projectId, tableId.getDataset())), tableId.getTable())));
             }
         }
-        catch (BigQueryException e) {
-            if (e.getCode() == 404 && e.getMessage().contains("Not found: Dataset")) {
+        catch (TrinoException e) {
+            if (e.getErrorCode() != BIGQUERY_LISTING_TABLE_ERROR.toErrorCode()) {
+                throw e;
+            }
+            BigQueryException bigQueryException = (BigQueryException) e.getCause();
```

Comment:
can this cast fail?

Reply:
It shouldn't since there's only one usage `BIGQUERY_LISTING_TABLE_ERROR` added in this PR and wraps a `BigQueryException`. I can add an `instanceof` check to be future proof?

Reply:
I think it'd be good to since it'll lead to runtime failures since we don't explicitly test for this failure.

Reply:
Fixed, actually cleaner with pattern matching.